examples/step-86/step-86.cc

```diff
@@ -0,0 +1,1387 @@
+
+    triangulation.execute_coarsening_and_refinement();
+    timer.stop();
+    pcout << "took (" << timer.cpu_time() << "s)" << std::endl << std::endl;
+  }
+
+  // As the surface is properly approximated by several adaptive steps, we may
+  // now distribute the degrees of
+  //  freedom to cells which are intersected by the discrete approximation.
+  template <int dim>
+  void LaplaceBeltramiSolver<dim>::distribute_activeDOFs()
+  {
+    pcout << "Distributing degrees of freedom... ";
+    Timer timer;
+    dof_handler.distribute_dofs(fe_collection);
+    locally_owned_dofs = dof_handler.locally_owned_dofs();
+    locally_relevant_dofs =
+      DoFTools::extract_locally_relevant_dofs(dof_handler);
+#ifdef USE_TRILINOS
+    completely_distributed_solution.reinit(dof_handler.locally_owned_dofs(),
+                                           MPI_COMM_WORLD);
+    locally_relevant_solution.reinit(locally_owned_dofs,
+                                     locally_relevant_dofs,
+                                     MPI_COMM_WORLD);
+    global_rhs.reinit(locally_owned_dofs, MPI_COMM_WORLD);
+#else
+    completely_distributed_solution.reinit(dof_handler.n_dofs());
+    global_rhs.reinit(dof_handler.n_dofs());
+#endif
+
+    timer.stop();
+    pcout << "took (" << timer.cpu_time() << "s)" << std::endl;
+  }
+
+  // Next, we initialize matrices for active DOFS and apply the constraints for
+  // the solution.
+  template <int dim>
+  void LaplaceBeltramiSolver<dim>::initialize_matrices()
+  {
+    pcout << "Initializing matrices... ";
+    Timer                  timer;
+    DynamicSparsityPattern dsp(dof_handler.n_dofs(), dof_handler.n_dofs());
+    constraints.clear();
+#ifdef USE_TRILINOS
+    constraints.reinit(locally_relevant_dofs);
+//    {
+//      // set one point to zero
+//      const unsigned int n_dofs_per_cell = fe_collection[0].dofs_per_cell;
+//      std::vector<types::global_dof_index> local_dof_indices(n_dofs_per_cell);
+//      for (const auto &cell : dof_handler.active_cell_iterators())
+//        if (cell->is_locally_owned() &&
+//            cell->active_fe_index() == ActiveFEIndex::lagrange)
+//          {
+//            cell->get_dof_indices(local_dof_indices);
+//            if (Utilities::MPI::this_mpi_process(MPI_COMM_WORLD) == 0)
+//              constraints.add_line(local_dof_indices[0]);
+//          }
+//    }
+#endif
+
+    DoFTools::make_hanging_node_constraints(dof_handler, constraints);
+    constraints.close();
+    DoFTools::make_sparsity_pattern(dof_handler, dsp, constraints);
+
+#ifdef USE_TRILINOS
+    SparsityTools::distribute_sparsity_pattern(dsp,
+                                               locally_owned_dofs,
+                                               MPI_COMM_WORLD,
+                                               locally_relevant_dofs);
+    global_matrix.reinit(locally_owned_dofs,
+                         locally_owned_dofs,
+                         dsp,
+                         MPI_COMM_WORLD);
```

Comment:
We typically do not use MPI_COMM_WORLD directly. Can you create a communicator ``mpi_comm`` inside the class and use that everywhere?